Initial commit for not showing UI w/edit profile b2c policy

jennyf19 opened the pull request:
- clean up unused stuff in b2cAuthority class
- Add UI test for edit profile
- add check for b2c authorityType in RequestBase to by-pass cache issue
- Add UIBehavior None
- Add b2c tests in desktop msal app and msal xamarin apps
- And some resharper suggestions

I have some more work to do, esp. in desktop app, but wanted to get something pushed for @jmprieur before the break.
[MSAL issue for context](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/588)

jennyf19 committed:
Initial commit for not showing UI w/edit profile b2c policy
- clean up unused stuff in b2cAuthority class
- Add UI test for edit profile
- add check for b2c authorityType in RequestBase to by-pass cache issue
- Add UIBehavior None
- Add b2c tests in desktop msal app and msal xamarin apps
- And some resharper suggestions

jennyf19 commented on msal/src/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs:
```diff
@@ -238,7 +239,7 @@ protected AuthenticationResult CacheTokenResponseAndCreateAuthenticationResult(M
             }
 
-            if (fromServer!= null && AuthenticationRequestParameters?.Account?.HomeAccountId != null)
+            if (fromServer!= null && AuthenticationRequestParameters?.Account?.HomeAccountId != null && AuthenticationRequestParameters.Authority.AuthorityType != AuthorityType.B2C)
```
`fromServer.UniqueObjectIdentifier` returns `"b9f....b2c_1_profileeditpolicy"`
and 
`AuthRequestParameters.Account.HomeObjectId` is `b9f...b2c_1_sisopolicy"` 
Using fiddler, we get a token back, so I added a check for b2c authorityType to skip this check and not get the `"Returned user identifiers do not match the sent user identifier"` error. 
Will look into this more next week.
 #Resolved

bgavrilMS replied:
Does the tenant id match - can we preserve that verification at least? 

In any case, this validation code is getting complex, please move it to a separate private method.  #Resolved

jennyf19 commented on msal/tests/Test.MSAL.NET.UIAutomation/MSALMobileTestHelper.cs:
```diff
@@ -309,5 +325,21 @@ public void PerformB2CSignInFlow(ITestController controller, LabUser user, B2CId
+            //Acquire token flow
+            controller.Tap(CoreUiTestConstants.AcquireTokenID);
+
+            controller.Tap(CoreUiTestConstants.B2CEditProfileContinueID, XamarinSelector.ByHtmlIdAttribute);
```
I guessed on the name of the id and got it. lol. 😹  #Resolved

bgavrilMS replied:
:D [insert thug life meme here] #Resolved

jennyf19 committed:
change to NoPrompt as per JM's comment

bgavrilMS commented on msal/tests/dev apps/DesktopTestApp/MainForm.cs:
```diff
@@ -498,5 +503,95 @@ private IEnumerable<string> SplitScopeString(string scopes)
+            {
+                ClearResultPageInfo();
+                string[] scopes = { "https://msidlabb2c.onmicrosoft.com/msidlabb2capi/read" };
+                string b2cClientId = "e3b9ad76-9763-4827-b088-80c7a7888f79";
```
I thought that clientIDs are read from the LabResponse? #Resolved

jennyf19 replied:
i fixed this. had to update to 4.6.2 to be compatible w/the lab. 

---
In reply to: [235680206](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1424#discussion_r235680206) [](ancestors = 235680206)

bgavrilMS commented on msal/tests/dev apps/DesktopTestApp/MainForm.cs:
```diff
@@ -498,5 +503,95 @@ private IEnumerable<string> SplitScopeString(string scopes)
+            {
+                ClearResultPageInfo();
+                string[] scopes = { "https://msidlabb2c.onmicrosoft.com/msidlabb2capi/read" };
+                string b2cClientId = "e3b9ad76-9763-4827-b088-80c7a7888f79";
```
client id should come from the lab reponse and be configured in the app. #Resolved

bgavrilMS commented on msal/src/Microsoft.Identity.Client/UIBehavior.cs:
```diff
@@ -61,6 +61,12 @@ struct UIBehavior
+        /// to a user who has previously acquired a token for a different policy. It is achieved
+        /// by sending <c>prompt=none</c> to the Azure AD service.
+        /// </summary>
+        public static readonly UIBehavior NoPrompt = new UIBehavior("none");
```
What are the implications of using this prompt in non-B2C scenarios? Is it something that we want to restrict users from doing, i.e. IF Authority != B2C and prompt == none THEN throw ex ? #Resolved

jennyf19 replied:
I had the same concern. @jmprieur - your thoughts on this?

---
In reply to: [235682552](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1424#discussion_r235682552) [](ancestors = 235682552)

bgavrilMS reviewed (changes requested):
Mostly questions about new prmpt behavior and client id in tests

henrik-me commented:
>             {

consider switch instead of if's #Resolved

---
Refers to: msal/tests/dev apps/XForms/XForms/SettingsPage.xaml.cs:109 in 506af75. [](commit_id = 506af7511140e852b1b8d2348b133c822965b4b3, deletion_comment = False)

henrik-me commented on msal/tests/Test.MSAL.NET.UIAutomation/MSALMobileTestHelper.cs:
```diff
@@ -309,5 +325,21 @@ public void PerformB2CSignInFlow(ITestController controller, LabUser user, B2CId
+            
+            controller.Tap(CoreUiTestConstants.AcquirePageID);
+            
+            SetUiBehavior(controller, CoreUiTestConstants.UIBehaviorNone);
```
>SetUiBehavior(controller, CoreUiTestConstants.UIBehaviorNone); [](start = 12, length = 62)

Makes sense to ask JM if this is how he envisions the setup. Also... what if the user is not logged in? Currently thinking the service side has to be part of deciding whether UI is shown or not. Would be great to get JM's perspective, and our friends in B2C (e.g. P and J). #Resolved

jennyf19 replied:
if it's used as a silent call, server side will return this:
 `One of two conditions was encountered: 1. The UiBehavior.Never flag was passed, but the constraint could not be honored, because user interaction was required. 2. An error occurred during a silent web authentication that prevented the http authentication flow from completing in a short enough time frame` 

---
In reply to: [236440625](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1424#discussion_r236440625) [](ancestors = 236440625)

henrik-me replied:
Trying to clarify my train of thoughts. 

Yeah, this flow doesn't make sense for the silent flow. My question is if the user is not logged in and the first flow you call is the edit profile flow... I assume regardless of us passing in none, the user should still be met with a prompt dialog... this is why I'm asking for server ack of what we are doing. If they treat none as do not show the auth dialogs, then... it will require the developer to know if the user is logged in or not. 

---
In reply to: [236450550](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1424#discussion_r236450550) [](ancestors = 236450550,236440625)

jennyf19 replied:
I see what you're saying. I thought you meant if the dev used prompt=none in another flow to mimic a silent call.
To answer your question, if the user is not signed in and does an edit profile w/prompt=none, we get this back from B2C: 
`- AADB2C90077: User does not have an existing session and request prompt parameter has a value of 'None'.`
if you do edit profile w/select account, you can select the account first (login) and then edit the profile.
so, yes, the dev will need to know if the user has logged in, or they can surface this after the user has logged in, I think that's how our samples are set up. You log-in first and then are given the option to edit profile.

---
In reply to: [236455405](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1424#discussion_r236455405) [](ancestors = 236455405,236450550,236440625)

henrik-me replied:
Great. As expected. It will though be hard as a developer to fully build the experience with-out keeping state separately... but perhaps that is good enough. We should have a wiki page explaining this (and the evo messages at time of writing keeping in mind that is all subject to change). Also would be good to investigate what the b2c service would prefer going forward that we send.

---
In reply to: [236458472](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1424#discussion_r236458472) [](ancestors = 236458472,236455405,236450550,236440625)

jennyf19 replied:
sounds good. i'll follow up w/b2c

---
In reply to: [236462577](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1424#discussion_r236462577) [](ancestors = 236462577,236458472,236455405,236450550,236440625)

henrik-me commented on msal/tests/Test.MSAL.NET.UIAutomation/XamarinDroidMSALTests.cs:
```diff
@@ -134,6 +134,19 @@ public void B2CFacebookProviderWithMicrosoftAuthorityAcquireTokenTest()
+        [Test]
+        public void B2CFacebookProviderEditPolicyAcquireTokenTest()
+        {
+            _msalMobileTestHelper.B2CFacebookProviderAcquireTokenSilentTest(xamarinController, LabUserHelper.GetLabResponseWithB2CFacebookProvider(), isB2CLoginAuthority:true);
```
>isB2CLoginAuthority:true [](start = 150, length = 24)

any reason you are specifying the name of the parameter? #Closed

jennyf19 replied:
just to make it clear what the boolean is for, but i can remove the name.

---
In reply to: [236440994](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1424#discussion_r236440994) [](ancestors = 236440994)

bgavrilMS replied:
Makes the code easier to read... #Pending

henrik-me replied:
yeah, imo not needed. The methods should be made for that purpose instead of having to add all the param names in the mehtod call... 

personal opinion: use consts for that instaed if needed.

---
In reply to: [237127305](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1424#discussion_r237127305) [](ancestors = 237127305)

henrik-me reviewed (changes requested):
:clock1:

jennyf19 committed:
address some pr comments

jennyf19 committed:
Merge remote-tracking branch 'origin/dev' into jennyf/b2cPolicy

jennyf19 committed:
change uibehavior name from NoPrompt to None

jennyf19 committed:
more changes
- remove unused clientids
- some resharper stuff
- no include uibehavior in auth request params

jennyf19 committed:
fix ui test failure
use live lab api instead of test now that the b2c changes have been pushed on their end.

jennyf19 committed:
one more...

jennyf19 committed:
pr comments
- use lab api for b2c client id in desktop app
- update desktop app to 4.6.2

bgavrilMS commented on msal/src/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs:
```diff
@@ -276,17 +258,45 @@ protected AuthenticationResult CacheTokenResponseAndCreateAuthenticationResult(M
 
+        private void ValidateAccountIdentifiers(ClientInfo fromServer)
+        {
+            if (fromServer == null || AuthenticationRequestParameters?.Account?.HomeAccountId == null ) return;
```
Prefer 
{
    return;
} #Resolved

bgavrilMS commented on msal/tests/dev apps/DesktopTestApp/MainForm.cs:
```diff
@@ -349,6 +356,11 @@ private UIBehavior GetUIBehavior()
 
+            if (noPrompt.Checked)
+            {
+                behavior = UIBehavior.NoPrompt; ;
```
double ; #Resolved

jennyf19 replied:
nice catch! thank you!

---
In reply to: [237127805](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1424#discussion_r237127805) [](ancestors = 237127805)

bgavrilMS commented on msal/tests/dev apps/DesktopTestApp/MainForm.cs:
```diff
@@ -63,6 +66,8 @@ public MainForm()
 
             LoadSettings();
             Logger.LogCallback = LogDelegate;
+            LabResponse labResponse = LabUserHelper.GetLabResponseWithB2CLocalAccountProvider();
```
We shouldn't make a web request to the lab for B2C everytime we open the app, because it is quite expensive. I believe that the way tests run today is that each test closes and reopens the app. So all our non-B2C tests make this call, which is not needed. 

Instead, you can initialize _b2CClientId = null, then when you click on B2CLogin go query the lab. If I press B2CLogin again, then _b2CClientId is initialized so we won't have to call the lab again. 

This is called lazy initialization, and it's perfectly ok to do it manually like I described. Otherwise, C# has a class for this called Lazy<T> https://docs.microsoft.com/en-us/dotnet/api/system.lazy-1?view=netframework-4.7.2 - we use it in a few places in MSAL.  #Resolved

jennyf19 replied:
awesome! thanks.

---
In reply to: [237133848](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1424#discussion_r237133848) [](ancestors = 237133848)

bgavrilMS commented on msal/tests/dev apps/DesktopTestApp/MainForm.cs:
```diff
@@ -491,12 +503,89 @@ private void cancelOperationButton_Click(object sender, EventArgs e)
+                ClearResultPageInfo();
+
+                _publicClientHandler.InteractiveAuthority = B2CAuthority;
+
```
I think you should also set the client ID #Resolved

bgavrilMS replied:
Yes, it is good to fetch it here, but I meant _publicClientHandler.ApplicationId = _b2CClientId;  #Resolved

bgavrilMS reviewed (changes requested):
Lazy initialisation for the B2C account, otherwise LGTM

jennyf19 committed:
Merge remote-tracking branch 'origin/dev' into jennyf/b2cPolicy

jennyf19 committed:
PR comments

jennyf19 committed:
one more.

jennyf19 committed:
Merge remote-tracking branch 'origin/dev' into jennyf/b2cPolicy

jennyf19 committed:
fix comment

henrik-me commented on msal/tests/dev apps/DesktopTestApp/MainForm.cs:
```diff
@@ -29,28 +29,31 @@
-        private CancellationTokenSource _cancellationTokenSource;
-
+        private const string PublicClientId = "0615b6ca-88d4-4884-8729-b178178f7c27";
+        private static string _b2CClientId = null;
```
>static [](start = 16, length = 6)

I don't think this needs to be static? #Closed

jennyf19 committed:
remove static

henrik-me commented on msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs:
```diff
@@ -285,6 +285,11 @@ private void CheckForDuplicateQueryParameters(string queryParams, IDictionary<st
+                return authorizationRequestParameters;
+            }
+
             authorizationRequestParameters[OAuth2Parameter.Prompt] = _uiBehavior.PromptValue;
```
this is a bit of a special scenario.... I think to best ensure that this doesn't break some time in the future (by someone not returning anymore....) 
I suggest that you invert the assignment and then only assign the prompt in case the uibehavior is not NoPrompt. makes sense? #Resolved

henrik-me commented:
>         public static LabResponse GetLabResponseWithB2CLocalAccountProvider()

nit: Consider removing the return type from all the methods... seems a bit too much information. #Resolved

---
Refers to: core/tests/Test.Microsoft.Identity.LabInfrastructure/LabUserHelper.cs:96 in 0952f29. [](commit_id = 0952f29d8b6e7265bbcdbb10748af85517284c0e, deletion_comment = False)

henrik-me commented on core/src/Instance/B2CAuthority.cs:
```diff
@@ -81,14 +80,6 @@ internal B2CAuthority(IValidatedAuthoritiesCache validatedAuthoritiesCache, stri
-
-            string canonicalAuthorityUri = string.Format(CultureInfo.InvariantCulture, MicrosoftOnline + b2cAuthority.AbsolutePath);
-            return canonicalAuthorityUri;
-        }
```
:) #Resolved

jennyf19 committed:
invert a thing

jennyf19 committed:
remove param name

jennyf19 committed:
henrik's random comment

henrik-me reviewed (approved):
:shipit:

jennyf19 committed:
merge conflict

jennyf19 committed:
fix request base.
